### PR TITLE
More OSL bridge refactors and improvements

### PR DESCRIPTION
**Pull request by est77**

### CODE REVIEW

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 1]
```diff
@@ -455,11 +685,52 @@ namespace
...
+
+        static void convert_closure(
+            CompositeSurfaceClosure&    composite_closure,
+            const Basis3d&              shading_basis,
+            const void*                 osl_params,
+            const Color3f&              weight)
+        {
+            const Params* p = reinterpret_cast<const Params*>(osl_params);
+
+            // Assume on of the mediums is air.
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: "one"
"media"

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 2]
```diff
@@ -539,10 +830,86 @@ namespace
...
+
+                values->m_weight = 1.0;
+                values->m_reflectance = Color3f(p->rd);
+                values->m_reflectance_multiplier = 1.0;
+                values->m_dmfp = luminance(Color3f(p->dmfp));
+                values->m_dmfp_multiplier = 1.0;
+                values->m_inside_ior = p->eta;
+                values->m_outside_ior = 1.0;
+#else
+                throw ExceptionOSLRuntimeError("Unknown subsurface profile ");
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: Missing profile name in exception

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 3]
```diff
@@ -539,10 +830,86 @@ namespace
...
+                {
+                    values =
+                        composite_closure.add_closure<DipoleBSSRDFInputValues>(
+                            SubsurfaceDirectionalDipoleID,
+                            shading_basis,
+                            weight,
+                            Vector3d(p->N));
+                }
+                else
+                    throw ExceptionOSLRuntimeError("Unknown subsurface profile");
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: Can we mention which profile is missing here as well?

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 4]
```diff
@@ -733,18 +1120,15 @@ InputValues* CompositeClosure::do_add_closure(
...
 
     assert(m_num_bytes + sizeof(InputValues) <= MaxPoolSize);
 
     // We use the luminance of the weight as the BSDF weight.
     const double w = luminance(weight);
-
-    // Ignore zero or negative weights.
-    if (w <= 0.0)
-        return 0;
+    assert (w > 0.0);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: Remove space before opening parenthesis.

Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: Is w going to always be strictly positive, regardless of user input? If not, then an assertion is not the right solution.

File: src/appleseed/renderer/kernel/shading/closures.h [hunk 5]
```diff
@@ -122,6 +122,20 @@ enum ClosureID
 };
 
 //
+// Exception class thrown in case of an OSL runtime error.
+//
+
+struct ExceptionOSLRuntimeError
+  : public foundation::Exception
+{
+    explicit ExceptionOSLRuntimeError(const char* what) :
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.h]: The colon goes on the next line:

    explicit ExceptionOSLRuntimeError(const char* what)
      : foundation::Exception(what)
    {
    }

File: src/appleseed/renderer/kernel/shading/closures.cpp (hunk 4 above)
Comment (est77) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: Yes. Closures with zero or negative weights are ignored. In this method w > 0.

**Comment by est77** [comment]
Fixed and rebased

**Pull request merged**